File: src/main/java/link/rdcn/user/AuthenticatedUser.java
package link.rdcn.user;

import scala.collection.immutable.Set;

/**
 * @Author renhao
 * @Description:
 * @Data 2025/6/24 11:02
 * @Modified By:
 */
public class AuthenticatedUser {
    private final String userId;
    private final Set<String> roles;
    private final Set<String> permissions;

    public AuthenticatedUser(String userId, Set<String> roles, Set<String> permissions) {
        this.userId = userId;
        this.roles = roles;
        this.permissions = permissions;
    }

    public String getUserId() {
        return userId;
    }

    public Set<String> getRoles() {
        return roles;
    }

    public Set<String> getPermissions() {
        return permissions;
    }
}

File: faird-core/src/main/scala/link/rdcn/user/AuthenticatedUser.scala
package link.rdcn.user

import link.rdcn.ConfigLoader
import link.rdcn.util.{CodecUtils, KeyBasedAuthUtils}
import org.json.JSONObject

import java.security.PublicKey

/**
 * @Author renhao
 * @Description:
 * @Data 2025/7/9 17:06
 * @Modified By:
 */
trait AuthenticatedUser

case class KeyAuthenticatedUser(
                                 serverId: String,
                                 nonce: String,
                                 issueTime: Long, //签发时间
                                 validTo: Long, //过期时间
                                 signature: Array[Byte] // UnionServer 私钥签名
                               ) extends AuthenticatedUser {
  def checkPermission(): Boolean = {
    val publicKey: Option[PublicKey] = ConfigLoader.fairdConfig.pubKeyMap.get(serverId)
    if (publicKey.isEmpty) false else {
      if (validTo > issueTime) {
        KeyBasedAuthUtils.verifySignature(publicKey.get, getChallenge(), signature) && System.currentTimeMillis() < validTo
      } else {
        KeyBasedAuthUtils.verifySignature(publicKey.get, getChallenge(), signature)
      }
    }
  }

  private def getChallenge(): Array[Byte] = {
    val jo = new JSONObject().put("serverId", serverId)
      .put("nonce", nonce)
      .put("issueTime", issueTime)
      .put("validTo", validTo)
    CodecUtils.encodeString(jo.toString)
  }
}
